shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs

```diff
@@ -36,97 +36,88 @@ public static string GetTypeDisplayName(object item, bool fullName = true)
-            var parts = genericFullName.Split('+');
-            var genericArguments = t.GetTypeInfo().GenericTypeArguments;
-            var index = 0;
-            var totalParts = parts.Length;
-            if (totalParts == 1)
-            {
-                var part = parts[0];
-                var num = part.IndexOf('`');
-                if (num == -1) return;
+            var genericArguments = type.GetGenericArguments();
+            ProcessNestedGenericType(builder, type, genericArguments, genericArguments.Length, fullName);
+        }
 
-                var name = part.Substring(0, num);
-                var numberOfGenericTypeArgs = int.Parse(part.Substring(num + 1));
-                sb.Append(fullName ? name : genericSimpleName.Substring(0, genericSimpleName.IndexOf('`')));
-                AppendGenericArguments(genericArguments, index, numberOfGenericTypeArgs, sb, fullName);
-                return;
+        private static void ProcessNestedGenericType(StringBuilder builder, Type type, Type[] genericArgs, int limit, bool fullName)
+        {
+            var offset = 0;
+            var nestingType = type.DeclaringType;
+
+            if (nestingType != null)
+            {
+                offset = nestingType.GetGenericArguments().Length;
             }
-            for (var i = 0; i < totalParts; i++)
+
+            if (fullName)
             {
-                var part = parts[i];
-                var num = part.IndexOf('`');
-                if (num != -1)
+                if (nestingType != null)
                 {
-                    var name = part.Substring(0, num);
-                    var numberOfGenericTypeArgs = int.Parse(part.Substring(num + 1));
-                    if (fullName || i == totalParts - 1)
-                    {
-                        sb.Append(name);
-                        AppendGenericArguments(genericArguments, index, numberOfGenericTypeArgs, sb, fullName);
-                    }
-                    if (fullName && i != totalParts - 1)
-                    {
-                        sb.Append("+");
-                    }
-                    index += numberOfGenericTypeArgs;
+                    ProcessNestedGenericType(builder, nestingType, genericArgs, offset, fullName);
+                    builder.Append('+');
                 }
                 else
                 {
-                    if (fullName || i == totalParts - 1)
-                    {
-                        sb.Append(part);
-                    }
-                    if (fullName && i != totalParts - 1)
-                    {
-                        sb.Append("+");
-                    }
+                    builder.Append(type.Namespace);
+                    builder.Append('.');
+                }
+            }
+
+            var genericPartIndex = -1;
+            if (type.IsGenericType)
+            {
+                genericPartIndex = type.Name.IndexOf('`');
+            }
+
+            if (genericPartIndex > 0)
```

Comment:
What is the case where genericPartIndex is not `> 0` for a generic type?
If type is generic then there should be `` ` `` present. So we don't need 2 ifs here.

Reply:
`Outer<int>.Inner` also count as generic type, but its name will be `Inner`, without  `` ` `` character.

Reply:
That is even better. Find backtick. If not found print type and return. If found, then we print appropriate name and generic args. Nomore need to check for generic type.